src/Chapter3.hs

```diff
@@ -374,6 +400,34 @@ after the fight. The battle has the following possible outcomes:
 
 -}
 
+{-
+Actually I don't see the point in writing separate
+Monster and Knight types as their properties are the same
+-}
+data Fighter = Fighter
+    { health :: Int
+    , attack :: Int
+    , gold   :: Int
+    } deriving (Show)
+
+{- Returns the updated knight, monster and fight result (-1 knight lost, 1 knight won, 0 draw) -}
+fight :: (Fighter, Fighter)-> (Fighter, Fighter, Int)
+fight (knight, monster)
+    | health knight  == 0 && health monster > 0 = (knight, monster, -1)
+    | health monster == 0 && health knight  > 0 = (knight { gold = gold knight + gold monster }, monster, 1)
+    | health monster == 0 && health knight == 0 = (knight, monster, 0)
+    | otherwise = fight (nextRound (knight, monster))
+    where
+        nextRound :: (Fighter, Fighter) -> (Fighter, Fighter)
+        nextRound (knight, monster) = (updatedKnight, updatedMonster)
+            where
+                updatedMonster :: Fighter
+                updatedKnight  :: Fighter
+                updatedMonster = monster { health = health monster - attack knight }
+                updatedKnight = if health monster > 0
+                                then knight { health = (health knight - attack monster) }
+                                else knight
```

Comment:
You can simplify it a bit and leave just `updateMonster` and `uodateKnight` functions:

 ```haskell
    | otherwise = fight (updatedKnight, updatedMonster)
  where
    updatedMonster :: Fighter
    updatedMonster = monster { health = health monster - attack knight }
            
    updatedKnight  :: Fighter
    updatedKnight = 
        if health monster > 0
        then knight { health = (health knight - attack monster) }
        else knight
```